libsolidity/formal/SMTEncoder.cpp

```diff
@@ -1406,37 +1408,56 @@ void SMTEncoder::bitwiseOperation(BinaryOperation const& _op)
 	auto commonType = _op.annotation().commonType;
 	solAssert(commonType, "");
 
+	auto [bvSize, isSigned] = typeBvSizeAndSigndness(commonType);
+
+	auto bvLeft = smtutil::Expression::int2bv(expr(_op.leftExpression(), commonType), bvSize);
+	auto bvRight = smtutil::Expression::int2bv(expr(_op.rightExpression(), commonType), bvSize);
+
+	optional<smtutil::Expression> result;
+	if (_op.getOperator() == Token::BitAnd)
+		result = bvLeft & bvRight;
+	else if (_op.getOperator() == Token::BitOr)
+		result = bvLeft | bvRight;
+	else if (_op.getOperator() == Token::BitXor)
+		result = bvLeft ^ bvRight;
+	else
+		solAssert(false, "");
+
+	if (result)
+		defineExpr(_op, smtutil::Expression::bv2int(*result, isSigned));
+}
+
+void SMTEncoder::bitwiseNotOperation(UnaryOperation const& _op)
+{
+	solAssert(_op.getOperator() == Token::BitNot, "");
+
+	auto [bvSize, isSigned] = typeBvSizeAndSigndness(_op.annotation().type);
+
+	auto bvOperand = smtutil::Expression::int2bv(expr(_op.subExpression(), _op.annotation().type), bvSize);
+	optional<smtutil::Expression> result = ~bvOperand;
+
+	if (result)
+		defineExpr(_op, smtutil::Expression::bv2int(*result, isSigned));
+}
+
+tuple<unsigned, bool> SMTEncoder::typeBvSizeAndSigndness(TypePointer const& _type)
+{
 	unsigned bvSize = 256;
 	bool isSigned = false;
-	if (auto const* intType = dynamic_cast<IntegerType const*>(commonType))
+	if (auto const* intType = dynamic_cast<IntegerType const*>(_type))
 	{
 		bvSize = intType->numBits();
 		isSigned = intType->isSigned();
 	}
-	else if (auto const* fixedType = dynamic_cast<FixedPointType const*>(commonType))
+	else if (auto const* fixedType = dynamic_cast<FixedPointType const*>(_type))
 	{
 		bvSize = fixedType->numBits();
 		isSigned = fixedType->isSigned();
 	}
-	else if (auto const* fixedBytesType = dynamic_cast<FixedBytesType const*>(commonType))
+	else if (auto const* fixedBytesType = dynamic_cast<FixedBytesType const*>(_type))
 		bvSize = fixedBytesType->numBytes() * 8;
 
-	auto bvLeft = smtutil::Expression::int2bv(expr(_op.leftExpression(), commonType), bvSize);
-	auto bvRight = smtutil::Expression::int2bv(expr(_op.rightExpression(), commonType), bvSize);
-
-	optional<smtutil::Expression> result;
-	if (_op.getOperator() == Token::BitAnd)
-		result = bvLeft & bvRight;
-	// TODO implement the other operators
-	else
-		m_errorReporter.warning(
-			1093_error,
-			_op.location(),
-			"Assertion checker does not yet implement this bitwise operator."
-		);
-
```

Comment:
Could add `else solAssert(false, "");` to ensure that this function is not called with non numeric types.

Reply:
I think now with `solAssert(result, "")` covers that case

Reply:
my bad, I'll add it to util function, as that is where it is now...